node/src/main/scala/com/wavesplatform/Importer.scala

```diff
@@ -188,58 +189,68 @@ object Importer extends ScorexLogging {
-            blockSize
-          } else {
-            ByteStreams.read(inputStream, blockBytes, 0, blockSize)
-          }
+      try {
+        val s1 = ByteStreams.read(inputStream, lenBytes, 0, Ints.BYTES)
+        if (s1 == Ints.BYTES) {
+          val blockSize = Ints.fromByteArray(lenBytes)
+
+          lazy val blockBytes = new Array[Byte](blockSize)
+          val factReadSize =
+            if (blocksToSkip > 0) {
+              // File IO optimization
+              ByteStreams.skipFully(inputStream, blockSize)
+              blockSize
+            } else {
+              ByteStreams.read(inputStream, blockBytes, 0, blockSize)
+            }
 
-        if (factReadSize == blockSize) {
-          if (blocksToSkip > 0) {
-            blocksToSkip -= 1
-          } else {
-            val blockV5 = blockchain.isFeatureActivated(
-              BlockchainFeatures.BlockV5,
-              blockchain.height + 1
-            )
-            val block =
-              (if (importOptions.format == Formats.Binary && !blockV5) Block.parseBytes(blockBytes)
-               else PBBlocks.vanilla(PBBlocks.addChainId(protobuf.block.PBBlock.parseFrom(blockBytes)), unsafe = true)).get
-            if (blockchain.lastBlockId.contains(block.header.reference)) {
-              Await.result(appendBlock(block).runAsyncLogErr, Duration.Inf) match {
-                case Left(ve) =>
-                  log.error(s"Error appending block: $ve")
-                  quit = true
-                case _ =>
-                  counter = counter + 1
-              }
+          if (factReadSize == blockSize) {
+            if (blocksToSkip > 0) {
+              blocksToSkip -= 1
             } else {
-              log.warn(s"Block $block is not a child of the last block ${blockchain.lastBlockId.get}")
+              val blockV5 = blockchain.isFeatureActivated(
+                BlockchainFeatures.BlockV5,
+                blockchain.height + 1
+              )
+              val block =
+                (if (importOptions.format == Formats.Binary && !blockV5) Block.parseBytes(blockBytes)
+                 else PBBlocks.vanilla(PBBlocks.addChainId(protobuf.block.PBBlock.parseFrom(blockBytes)), unsafe = true)).get
+              if (blockchain.lastBlockId.contains(block.header.reference)) {
+                Await.result(appendBlock(block).runAsyncLogErr, Duration.Inf) match {
+                  case Left(ve) =>
+                    log.error(s"Error appending block: $ve")
+                    quit = true
+                  case _ =>
+                    counter = counter + 1
+                }
+              } else {
+                log.warn(s"Block $block is not a child of the last block ${blockchain.lastBlockId.get}")
+              }
             }
+          } else {
+            log.info(s"$factReadSize != expected $blockSize")
+            log.info(s"reestablishing input stream")
+            inputStream.close()
+            inputStream = getInputStream()
```

Comment:
Just a notice: actually we have `blockSize` for each block, so we don't need to read whole `BlockInfoAtHeight` sequence to find the offset.